Creating TFRecords example - generate TFRecords from data samples

dimitreOliveira opened the pull request:
Hello, this is my first small contribution to the project, I have been using the framework for quite some time, and I 
hope this can be useful for other users as well.

Using TFRecords has many advantages, and I have been using them for a while whenever I have the chance.
With this code example, I hope that other developers can have a clean script as a starting point to generate TFRecords from their datasets.

The goal of this code example is not to demonstrate how to convert the COCO dataset to TFRecords, its goal is to demonstrate how to generate TFRecords using data that has many different data types, this way the developers can learn from a clean but yet generic example, that can easily translate to other domains.

Please let me know if I need to change something, I hope I can contribute more times in the future.

fchollet commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,193 @@
+## Downloading the COCO2017 dataset
+
+We will be using the [COCO2017](https://cocodataset.org/) dataset because it has many
+different types of features and will serve as a good example of how to encode different
```
We should enumerate the features and their properties

dimitreOliveira replied:
Added information about the dataset and its features.

fchollet commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,193 @@
+pprint.pprint(annotations[60])
+
+"""
+## Parameters
```
We should add some explanatory text

dimitreOliveira replied:
Text added with more details.

dimitreOliveira replied:
Moved this section to be right after the dataset features description, this way they are complementary.

fchollet commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,193 @@
+    plt.show()
+
+"""
+## Conclusion
```
We're missing a section about training a model with the dataset. E.g. turning this into a dataset that yields (features, labels).

dimitreOliveira replied:
Added one section for the training model, I made it very simple, to avoid taking attention from the TFRecords part, and to keep the script smaller.

fchollet commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,193 @@
+
+Now instead of reading images and annotations from different folders, we can have all the
+information of the samples reading from the same source, the TFRecords that we just
+created. This process makes storing and reading the data much more easy and efficient.
```
We may also want to talk about storing the data remotely (like on Google Cloud) and loading/buffering the data for training from the remote location without batch-downloading it. This is a relevant use case.

dimitreOliveira replied:
I have added this use case to the introduction, so it can complement the TPU example.

fchollet reviewed (commented):
Thanks for the PR! While this is not directly Keras-related, this is very useful info. Adding a section about training a Keras model would complete the example.

In general we need more detailed explanations throughout.

dimitreOliveira commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,193 @@
+
+## Introduction
+
+The TFRecord format is a simple format for storing a sequence of binary records.
```
Introduction updated with more information.

dimitreOliveira commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,193 @@
+
+The TFRecord format is a simple format for storing a sequence of binary records.
+
+Converting your data into TFRecords has many advantages, especially if later, you load it
```
Introduction updated with more information.

dimitreOliveira commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,193 @@
+    os.makedirs(tfrecords_dir)  # Creating output folder
+
+"""
+## Auxiliary functions
```
Text added with more details.

dimitreOliveira commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,193 @@
+def _bytes_feature(value):
+    """Returns a bytes_list from a string / byte."""
+    if isinstance(value, type(tf.constant(0))):
+        value = value.numpy()  # BytesList won't unpack a string from an EagerTensor.
```
Functions refactored and added one specific for images.

dimitreOliveira commented:
Hi @fchollet thank you very much for your extensive review. 

This is turning to be a learning experience, even though I am the one adding the example :-)

About your comments:
- Introduction: I have added more information to make it more clear why/when one should use TFRecords, let me know if it is enough now, I also have improved the TFRecord use case.
- Points with the need for more explanatory text: I have added more text to make things clearer.
- COCO dataset features, I have added more information about the dataset and its features.
- Added one function to serialize images, and fixed the string function.
- Added one section for the training model, I made it very simple, to avoid taking attention from the TFRecords part, and to keep the script smaller, but let me know if you want a more complete code for this training section,

fchollet commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,291 @@
+`n_tfrecods` is total number of TFRecords that we will create.
+"""
+
+n_samples = 4096
```
For consistency with other examples, use `num_` as the number prefix (everywhere).

fchollet commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,291 @@
+model.compile(
+    optimizer=tf.keras.optimizers.Adam(),
+    loss=tf.keras.losses.SparseCategoricalCrossentropy(),
+    metrics=["sparse_categorical_accuracy"],
```
Since you're passing the optimizer and loss as objects, for consistency also pass the metrics as objects

fchollet commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,291 @@
+
+train_filenames = tf.io.gfile.glob(tfrecords_dir + "/*.tfrec")
+batch_size = 64
+epochs = 5
```
You can set this to 1 since we're not training for real

fchollet commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,291 @@
+
+def get_dataset(filenames, batch_size):
+    """
+        Return a Tensorflow dataset ready for training or inference.
```
"For training or evaluation". This is a labeled dataset, so it's not for inference.

fchollet commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,291 @@
+    """
+    dataset = (
+        tf.data.TFRecordDataset(
+            filenames, num_parallel_reads=tf.data.experimental.AUTOTUNE
```
Prefer importing the constant "AUTOTUNE" just once and then reuse it (shorter)

fchollet commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,291 @@
+"""
+## Training a simple model with the generated TFRecords
+
+Explain that you are able to do not use all features from the dataset
```
TODO?

dimitreOliveira replied:
Yep, forgot to add this information, sorry.

fchollet reviewed (commented):
Thank you for the update! It looks good! I added a few edits, please pull them first.

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+by contrast, [COCO](https://cocodataset.org/) stores data into two folders,
+"images" and "annotations").
+
+An important use case of TFRecords is training on TPUs: TPUs are fast enough that
+they usually benefit from optimized I/O operations. In addition, TPUs require
+data to be stored remotely (e.g. on Google Cloud Storage), and storing your data as TFRecords
+makes it easier to the load data without batch-downloading it first.
```
Suggestion:

- Minor grammatical stuff
- Use `TFRecord` without the "s" (https://www.tensorflow.org/tutorials/load_data/tfrecord?hl=fr-EN)


```suggestion
An important use case of the TFRecord data format  is training on TPUs. First, TPUs are fast
enough to benefit from optimized I/O operations. In addition, TPUs require
data to be stored remotely (e.g. on Google Cloud Storage) and using the TFRecord format
makes it easier to load the data without batch-downloading.
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+data to be stored remotely (e.g. on Google Cloud Storage), and storing your data as TFRecords
+makes it easier to the load data without batch-downloading it first.
+
+Performance with TFRecords can be further improved if you use it with the
+[tf.data](https://www.tensorflow.org/guide/data) API.
```
Suggestion:

```suggestion
Performance using the TFRecord format can be further improved if you also use combine 
it with the [tf.data](https://www.tensorflow.org/guide/data) API.
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+import matplotlib.pyplot as plt
+
+"""
+## Downloading the data
```
Suggestion

- Specify the dataset name used in the example

```suggestion
## Download the COCO2017 dataset
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+    "http://images.cocodataset.org/annotations/annotations_trainval2017.zip"
+)
+
+# Download image files
```
Suggestion:

```suggestion
# Download the image files
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+    )
+    os.remove(image_zip)
+
+# Download caption annotation files
```
Suggestion:

```suggestion
# Download the caption annotation files
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+    )
+    os.remove(annotation_zip)
+
+print("Dataset is downloaded and extracted successfully.")
```
Suggestion:
```suggestion
print("The COCO dataset has been downloaded and extracted successfully.")
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+"""
+## The COCO2017 dataset
+
+We will be using the [COCO2017](https://cocodataset.org/) dataset, because it has many
```
Question: should this introduction to why you're using COCO be somewhere before the "downloading the dataset" section?

dimitreOliveira replied:
Hi @8bitmp3 , I left this introduction after downloading the data because I wanted to follow this introduction with a display of some annotation samples, so I needed it after the data is downloaded, but alternatively, I could put this introduction before the download section and put this sample as text instead of printing it, what do you think?

fchollet replied:
IMO you can put the first paragraph here under "Download the COCO2017 dataset" and keep the rest here. The title of this section can then be "Contents of the COCO2017 dataset"

dimitreOliveira replied:
Sounds good!

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+## The COCO2017 dataset
+
+We will be using the [COCO2017](https://cocodataset.org/) dataset, because it has many
+different types of features (images, floating point data, lists...):
```
Suggestion:

```suggestion
different types of features, including images, floating point data, and lists.
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+
+We will be using the [COCO2017](https://cocodataset.org/) dataset, because it has many
+different types of features (images, floating point data, lists...):
+it will serve as a good example of how to encode different features as TFRecords.
```
Suggestion:

- Use `TFRecord` without the "s" (https://www.tensorflow.org/tutorials/load_data/tfrecord?hl=fr-EN)

```suggestion
It will serve as a good example of how to encode different features into the TFRecord 
format.
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+different types of features (images, floating point data, lists...):
+it will serve as a good example of how to encode different features as TFRecords.
+
+This dataset has two sets of fields: images, and annotation meta-data.
```
Suggestion:

```suggestion
This dataset has two sets of fields: images and annotation meta-data.
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+
+This dataset has two sets of fields: images, and annotation meta-data.
+
+Images are a collection of ".jpg" files, and the meta-data is a ".json" file which,
```
Suggestion:

```suggestion
The images are a collection of JPG files and the meta-data are stored in a JSON file which,
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
````diff
@@ -0,0 +1,290 @@
+iscrowd: 0 or 1, is single object or a collection
+```
+
+### Looking at the annotations of a single data sample
````
Suggestion:

```suggestion
### Explore the annotations of a single data sample
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+    os.makedirs(tfrecords_dir)  # creating TFRecords output folder
+
+"""
+## TFRecords functions
```
Suggestion:

```suggestion
## Define TFRecords helper functions
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+
+
+"""
+## Generating TFRecords
```
Suggestion:
- Use `TFRecord` without the "s" (https://www.tensorflow.org/tutorials/load_data/tfrecord?hl=fr-EN)

```suggestion
## Generate data in the TFRecord format
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+"""
+## Generating TFRecords
+
+The generated TFRecrods will have the file name format `file_{number}.tfrec`. This is
+optional, but writing the number of files at the file name can make counting easier.
```
Suggestion:

- Minor refactoring
- Fixing misspelling ("TFRecord")

```suggestion
Let's generate the COCO2017 data in the TFRecord format. The format will be
`file_{number}.tfrec` (this is optional, but including the number sequences in the file
names can make counting easier).
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+            writer.write(example.SerializeToString())
+
+"""
+## Looking at a generated TFRecord of a single data sample
```
Suggestion:

```suggestion
## Explore one sample from the generated TFRecord
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+    plt.show()
+
+"""
+## Training a simple model with the generated TFRecords
```
Suggestion:

```suggestion
## Train a simple model using the generated TFRecords
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+"""
+
+"""
+## Dataset functions
```
Suggestion:

```suggestion
## Define dataset helper functions
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+"""
+## Conclusion
+
+Now, instead of reading images and annotations from different folders, we can have all the
+information we need from a single source, the TFRecords that we just
+created. This process makes storing and reading the data simpler and more efficient.
```
Suggestion:

- Minor refactoring
- Add one more mention of the the TFRecord and tf.train.Example tutorial on tf.org

```suggestion
This example demonstrates that instead of reading images and annotations from different sources
you can have your data coming from a single source thanks to TFRecord. This process can
make storing and reading data simpler and more efficient. For more information, you can go 
to the [TFRecord and tf.train.Example](https://www.tensorflow.org/tutorials/load_data/tfrecord) tutorial.
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+Author: [Dimitre Oliveira](https://www.linkedin.com/in/dimitre-oliveira-7a1a0113a/)
+Date created: 2021/02/25
+Last modified: 2021/02/25
+Description: Converting data to TFRecords files.
```
Suggestion:

TFRecord (without the "s") (https://www.tensorflow.org/tutorials/load_data/tfrecord?hl=fr-EN)

```suggestion
Description: Converting data to the TFRecord format.
```

8bitmp3 commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+[tf.data](https://www.tensorflow.org/guide/data) API.
+
+Here, we will learn how to convert data of different types (image, text, and numeric) into
+TFRecords.
```
Suggestion:

- Use `TFRecord` without the "s" (https://www.tensorflow.org/tutorials/load_data/tfrecord?hl=fr-EN)

```suggestion
TFRecord.
```

8bitmp3 reviewed (commented):
Hi @dimitreOliveira @fchollet I've suggested a few suggestions and fixes. Please note that it's `TFRecord` (singular, so there's no "s") (https://www.tensorflow.org/tutorials/load_data/tfrecord?hl=fr-EN). Hope this helps 👍

dimitreOliveira commented on examples/keras_recipes/creating_tfrecords.py:
```diff
@@ -0,0 +1,290 @@
+Performance with TFRecords can be further improved if you use it with the
+[tf.data](https://www.tensorflow.org/guide/data) API.
+
+Here, we will learn how to convert data of different types (image, text, and numeric) into
```
Hi @8bitmp3 , by numeric, I am referring to data with integer and float type.

dimitreOliveira commented:
Hello @fchollet and @8bitmp3 , thank you both for the extensive and detailed review, the code is looking much better now!

I have added the suggested changes, with the exception of one comment from @8bitmp3 about moving an informational section, but I have added a comment answering his request, please let me know what you think.

fchollet reviewed (commented):
Looks good, after the final edit, please add the autogenerated files, then we can merge it. Thanks for the great contribution!

dimitreOliveira commented:
You're welcome @fchollet the pleasure was mine.

I have made the final edit you suggested about the text section describing the dataset, and have also added the autogenerated files as you asked, it is a little hard to validate the consistency of the .ipynb and .md files, but by what I saw, it seems to be ok, let me know if you need anything else.

fchollet reviewed (approved):
Thank you!
